lib/plugindata/cas.go

```diff
@@ -0,0 +1,166 @@
+	UpdatePluginData(context.Context, types.PluginDataUpdateParams) error
+}
+
+// CompareAndSwap represents modifier struct
+type CompareAndSwap[T any] struct {
+	client      Client
+	name        string
+	kind        string
+	backoffBase time.Duration
+	backoffMax  time.Duration
+	encode      func(T) map[string]string
+	decode      func(map[string]string) T
+}
+
+// NewCAS returns modifier struct
+func NewCAS[T any](
+	client Client, name,
+	kind string,
+	encode func(T) map[string]string,
+	decode func(map[string]string) T,
+) *CompareAndSwap[T] {
+	return &CompareAndSwap[T]{
+		client,
+		name,
+		kind,
+		backoffBase,
+		backoffMax,
+		encode,
+		decode,
+	}
+}
+
+// Create tries to perform compare-and-swap update of a plugin data assuming that it does not exist
+//
+// fn callback function receives current plugin data value and returns modified value and
+// error.
+//
+// Please note that fn might be called several times due to CAS backoff, hence, you must be careful
+// with things like I/O ops and channels.
+func (c *CompareAndSwap[T]) Create(
+	ctx context.Context,
+	resource string,
+	cb func() (T, error),
+) (T, error) {
+	emptyData := *new(T)
+
+	existingData, err := c.getPluginData(ctx, resource)
+	if err != nil && !trace.IsNotFound(err) {
+		return emptyData, trace.Wrap(err)
+	}
+
+	if existingData != nil {
+		return emptyData, trace.AlreadyExists("plugin data already exists")
+	}
+
+	modifiedData, err := cb()
+	if err != nil {
+		return emptyData, trace.Wrap(err, "callback failed to update PluginData")
+	}
+
+	err = c.updatePluginData(ctx, resource, modifiedData, emptyData)
+	if err == nil {
+		return modifiedData, nil
+	}
+
+	return emptyData, trace.Wrap(err)
+}
+
+// Update tries to perform compare-and-swap update of a plugin data assuming that it exist
+//
+// fn callback function receives current plugin data value and returns modified value and
+// error. If fn has failed to update data due to a diff, it must return CompareFailed error.
+func (c *CompareAndSwap[T]) Update(
+	ctx context.Context,
+	resource string,
+	cb func(T) (T, error),
+) (T, error) {
+	emptyData := *new(T)
+
+	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
+	for {
+		oldData, err := c.getPluginData(ctx, resource)
+		if err != nil {
+			return emptyData, trace.Wrap(err)
+		}
+
+		if oldData == nil {
+			return emptyData, trace.NotFound("Plugin data does not exist")
+		}
+
+		cbData := *oldData
+		expectData := *oldData
+
+		newData, err := cb(cbData)
+		if err != nil && !trace.IsCompareFailed(err) {
+			return emptyData, trace.CompareFailed("Callback failed to update PluginData")
+		}
+
+		err = c.updatePluginData(ctx, resource, newData, expectData)
+		if err == nil {
+			return newData, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return emptyData, trace.Wrap(err)
+		}
+		if err := backoff.Do(ctx); err != nil {
+			return emptyData, trace.Wrap(err)
+		}
+	}
+}
+
+// TODO: Implement Upsert method
```

Comment:
```suggestion
// Upsert tries to perform compare-and-swap update of a plugin data.
func (c *CompareAndSwap[T]) Upsert(ctx context.Context, resource string, upsertData T) (T, error) {
	emptyData := *new(T)

	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
	for {
		existingData, err := c.getPluginData(ctx, resource)
		if trace.IsNotFound(err) {
			existingData := emptyData
		} else if err != nil {
			return emptyData, trace.Wrap(err)
		}

		err = c.updatePluginData(ctx, resource, upsertData, existingData)
		if err == nil {
			return upsertData, nil
		} else if !trace.IsCompareFailed(err) {
			return emptyData, trace.Wrap(err)
		} else if err := backoff.Do(ctx); err != nil {
			return emptyData, trace.Wrap(err)
		}
	}
}```

Reply:
Not exactly :-) We need callback and backoff because it is an extension to the update method. Let's follow YAGNI here: I update comment with "when it is required". As we are not using it right now, I do not see a good reason to add it.